### PR TITLE
refactor to use jwt instead of spotify token

### PR DESCRIPTION
**Pull request by dev-rifaii**

### CODE REVIEW

File: src/main/java/perosnal/spotifymatcher/controller/AuthenticationController.java [hunk 1]
```diff
@@ -19,6 +20,8 @@ public class AuthenticationController {
     private final SpotifyApiService spotifyApiService;
     private final SpotifyAuthorization spotifyAuthorization;
 
+    private final JwtTokenValidator jwtTokenValidator;
```
Comment (Jezorko) [inline src/main/java/perosnal/spotifymatcher/controller/AuthenticationController.java:23]: Never used here

File: src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    public void validateJwt(String jwt) {
+        Jwts.parserBuilder()
+                .setSigningKey(publicKey())
+                .build()
+                .parseClaimsJws(jwt)
+                .getBody();
+    }
+
+    @SneakyThrows
+    public PublicKey publicKey() {
```
Comment (Jezorko) [inline src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java:40]: Method may be `private`:
```suggestion
    private PublicKey publicKey() {
```

File: src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        KeyFactory kf = KeyFactory.getInstance("RSA");
+        PublicKey pk = kf.generatePublic(spec);
+        return pk;
+    }
+
+    public String getUserSpotifyId(String jwt) {
+        Jws<Claims> claims = Jwts.parserBuilder()
+                .setSigningKey(publicKey())
+                .build()
+                .parseClaimsJws(jwt);
```
Comment (Jezorko) [inline src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java:58]: Duplicate code should be extracted to a new method.

File: src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        System.out.println(claims.getBody().get("user_id"));
```
Comment (Jezorko) [inline src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java:59]: Either use a logger (`@Slf4j`) or remove this print statement entirely.
```suggestion
```

File: src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+        return claims.getBody().get("user_id").toString();
```
Comment (Jezorko) [inline src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java:61]: This method should return `Optional` as `user_id` may not be present in the token.

File: src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        Jwts.parserBuilder()
+                .setSigningKey(publicKey())
+                .build()
+                .parseClaimsJws(jwt)
+                .getBody();
+    }
+
+    @SneakyThrows
+    public PublicKey publicKey() {
+        String rsaPublicKey = httpRequestSender.genericRequest(AUTH_SERVER_URL+"/api/key");
```
Comment (Jezorko) [inline src/main/java/perosnal/spotifymatcher/security/JwtTokenValidator.java:41]: The public key should be cached somewhere so we don't call the auth server every time we need it.

File: src/main/java/perosnal/spotifymatcher/service/UserService.java [hunk 7]
```diff
@@ -90,15 +96,14 @@ public List<User> getMatches(String accessToken) {
...
-
-    public List<String> getTopTracks(String accessToken) {
-        return userRepository.getById(spotifyApiService.getIdByToken(accessToken))
-                .getTracks();
+    public List<Track> getTracksDetails(String jwt) {
+        jwtTokenValidator.validateJwt(jwt);
+        User user = userRepository.getById(jwtTokenValidator.getUserSpotifyId(jwt));
+        List<Track> tracks = new ArrayList<>();
+        for (String id : user.getTracks()) {
+            tracks.add(trackRepository.getById(id));
```
Comment (Jezorko) [inline src/main/java/perosnal/spotifymatcher/service/UserService.java:104]: Tracks from repository should be fetched with a batch query, and not individually.

**Review by Jezorko — CHANGES_REQUESTED** [review]
Please add more tests (:

File: src/main/java/perosnal/spotifymatcher/model/Track.java [hunk 8]
```diff
@@ -1,13 +1,24 @@
...
+import javax.persistence.Table;
+
+@Data
 @Builder
-@Getter
-@ToString
+@Entity
+@AllArgsConstructor
+@NoArgsConstructor
+@JsonIgnoreProperties({"hibernateLazyInitializer"})
```
[the remark this answers is not in the document]
Comment (dev-rifaii) [reply src/main/java/perosnal/spotifymatcher/model/Track.java:18]: All args constructor is for @Builder but not sure why when I remove @NoArgsConstructor I get:
Class 'Track' should have [public, protected] no-arg constructor

**Pull request merged**